astropy/table/index.py

```diff
@@ -0,0 +1,781 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+from copy import deepcopy
+import numpy as np
+
+from ..extern import six
+from .bst import BST, FastBST, FastRBT, MinValue, MaxValue
+from .sorted_array import SortedArray
+from ..time import Time
+
+'''
+The Index class can use several implementations as its
+engine. Any implementation should implement the following:
+
+__init__(data, row_index) : initialize index based on key/row list pairs
+add(key, row) -> None : add (key, row) to existing data
+remove(key, data=None) -> boolean : remove data from self[key], or all of
+                                    self[key] if data is None
+shift_left(row) -> None : decrement row numbers after row
+shift_right(row) -> None : increase row numbers >= row
+find(key) -> list : list of rows corresponding to key
+range(lower, upper, bounds) -> list : rows in self[k] where k is between
+                               lower and upper (<= or < based on bounds)
+sort() -> list of rows in sorted order (by key)
+replace_rows(row_map) -> None : replace row numbers based on slice
+items() -> list of tuples of the form (key, data)
+
+Notes
+-----
+    When a Table is initialized from another Table, indices are
+    (deep) copied and their columns are set to the columns of the new Table.
+
+    Column creation:
+    Column(c) -> deep copy of indices
+    c[[1, 2]] -> deep copy and reordering of indices
+    c[1:2] -> reference
+    array.view(Column) -> no indices
+'''
+
+class QueryError(ValueError):
+    '''
+    Indicates that a given index cannot handle the supplied query.
+    '''
+    pass
+
+class Index(object):
+    '''
+    The Index class makes it possible to maintain indices
+    on columns of a Table, so that column values can be queried
+    quickly and efficiently. Column values are stored in lexicographic
+    sorted order, which allows for binary searching in O(log n).
+
+    Parameters
+    ----------
+    columns : list or None
+        List of columns on which to create an index. If None,
+        create an empty index for purposes of deep copying.
+    engine : type or None
+        Indexing engine class to use, from among SortedArray, BST,
+        FastBST, and FastRBT. If the supplied argument is None (by
+        default), use SortedArray.
+    data : SortedArray, BST, FastBST, FastRBT, or None
+        Engine data to copy
+    '''
+    def __new__(cls, *args, **kwargs):
+        self = super(Index, cls).__new__(cls)
+        self.__init__(*args, **kwargs)
+        return SlicedIndex(self, slice(0, 0, None), original=True)
+
+    def __init__(self, columns, engine=None, data=None):
+        from .table import Table
+
+        if data is not None: # create from data
```

Comment:
Since in this case the engine is pre-determined anyways, it seems to me we could do away with the separate `data` argument (I can't find any code that uses it yet?), and allow `engine` to accept a type or an instance.  If `not isinstance(engine, type)` then we would do the same as in this clause, and otherwise instantiate the engine with the row data like below.

On the other hand, I like the symmetry of this with the attributes of the same names.  It just seems a little confusing to me (especially considering that you can also pass in something completely different as `engine`, but it's ignored if `data is not None`).